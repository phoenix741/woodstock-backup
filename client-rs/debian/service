[Unit]
Description=Woodstock Client Daemon

[Service]
ExecStart=/usr/bin/ws_client_daemon

[Install]
WantedBy=network.target