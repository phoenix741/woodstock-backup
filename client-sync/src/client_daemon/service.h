#pragma once

#include <QtService>

#include "woodstockclient.h"

class ClientService : public QtService::Service
{
    Q_OBJECT
public:
    explicit ClientService(int &argc, char **argv);

protected:
    virtual ClientService::CommandResult onStart() override;
    virtual ClientService::CommandResult onStop(int &exitCode) override;

private:
    std::unique_ptr<WoodstockClient> m_client;
};
