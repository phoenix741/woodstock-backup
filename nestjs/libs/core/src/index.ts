export * from './event-emitter';
export * from './commands';
export * from './config';
export * from './core.module';
export * from './input-output';
export * from './utils';
export * from './core.constant';
