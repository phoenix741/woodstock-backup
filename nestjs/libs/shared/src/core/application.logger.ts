import { Injectable, LoggerService } from '@nestjs/common';
import * as logform from 'logform';
import * as mkdirp from 'mkdirp';
import { join } from 'path';
import { createLogger, format, Logger, transports } from 'winston';
import 'winston-daily-rotate-file';

const { combine, timestamp, printf, colorize } = format;

const applicationFormat = printf((info: logform.TransformableInfo) => {
  return `${info.timestamp} [${(info.context || '').padEnd(25, ' ')}] ${info.level}: ${info.message}`;
});

@Injectable()
export class ApplicationLogger implements LoggerService {
  #logger: Logger;

  constructor(readonly worker?: string, readonly console = true) {
    const logPath = join(process.env.BACKUP_PATH || '', 'log');
    mkdirp(logPath);

    this.#logger = createLogger({
      level: process.env.LOG_LEVEL || 'info',
      format: combine(timestamp(), applicationFormat),
      transports: [
        ...(console
          ? [
              new transports.Console({
                format: combine(colorize(), timestamp(), applicationFormat),
              }),
            ]
          : []),
        new transports.DailyRotateFile({
          filename: join(logPath, `application-${worker}-%DATE%.log`),
          datePattern: 'YYYY-MM-DD',
          zippedArchive: true,
          maxSize: '2m', // Config
          maxFiles: '31d', // Config
          createSymlink: true,
          symlinkName: `application-${worker}.log`,
        }),
      ],
      exceptionHandlers: [
        new transports.Console({
          format: combine(colorize(), timestamp(), applicationFormat),
        }),
        new transports.DailyRotateFile({
          filename: join(logPath, `exceptions-${worker}-%DATE%.log`),
          datePattern: 'YYYY-MM-DD',
          zippedArchive: true,
          maxSize: '2m', // Config
          maxFiles: '31d', // Config
          createSymlink: true,
          symlinkName: `exceptions-${worker}.log`,
        }),
      ],
    });
  }

  log(message: string | Record<string, unknown>, context?: string): void {
    this.#logger.info(typeof message === 'string' ? { context, message } : { context, ...message });
  }

  error(message: string | Record<string, unknown>, trace?: string, context?: string): void {
    this.#logger.error(typeof message === 'string' ? { context, trace, message } : { context, trace, ...message });
  }

  warn(message: string | Record<string, unknown>, context?: string): void {
    this.#logger.warn(typeof message === 'string' ? { context, message } : { context, ...message });
  }

  debug(message: string | Record<string, unknown>, context?: string): void {
    this.#logger.debug(typeof message === 'string' ? { context, message } : { context, ...message });
  }

  verbose(message: string | Record<string, unknown>, context?: string): void {
    this.#logger.verbose(typeof message === 'string' ? { context, message } : { context, ...message });
  }
}
