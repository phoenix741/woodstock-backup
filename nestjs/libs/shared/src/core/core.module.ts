import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { CommandsModule } from '../commands';
import { ApplicationConfigModule } from '../config';
import { FilesModule } from '../files';
import { InputOutputModule } from '../input-output';
import { ManifestModule } from '../manifest';
import { PoolModule } from '../pool';
import { RefcntModule } from '../refcnt';
import { ScannerModule } from '../scanner';
import { StatisticsModule } from '../statistics';

export const MODULES = [
  ApplicationConfigModule,
  CommandsModule,
  ConfigModule,
  FilesModule,
  InputOutputModule,
  ManifestModule,
  PoolModule,
  RefcntModule,
  ScannerModule,
  StatisticsModule,
];

@Module({
  imports: [...MODULES],
  providers: [],
  exports: [...MODULES],
})
export class CoreModule {}
