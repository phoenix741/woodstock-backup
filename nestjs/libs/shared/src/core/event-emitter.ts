import * as EventEmitter from 'events';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
EventEmitter.defaultMaxListeners = 20;
