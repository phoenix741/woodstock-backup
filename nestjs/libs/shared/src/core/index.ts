export * from './event-emitter';
export * from './core.module';
export * from './application.logger';
