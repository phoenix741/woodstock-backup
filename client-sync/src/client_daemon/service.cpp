#include "service.h"

ClientService::ClientService(int &argc, char **argv) : Service{argc, argv}, m_client(new WoodstockClient())
{
}

ClientService::CommandResult ClientService::onStart()
{
    qDebug() << "Service was started";

    m_client->listen();

    qDebug() << "Service is listening";
    return ClientService::CommandResult::Completed; //service is now assumed started
}

ClientService::CommandResult ClientService::onStop(int &exitCode)
{
    qDebug() << "Stop received";
    
    exitCode = 0;

    m_client->stop();

    qDebug() << "Stopped";
    return ClientService::CommandResult::Completed;
}
