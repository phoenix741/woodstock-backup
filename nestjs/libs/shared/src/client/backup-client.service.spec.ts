import { NotFoundException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { CoreModule, ProtobufService } from '@woodstock/core';
import { count, from, toArray } from 'ix/asynciterable';
import * as Long from 'long';
import { join } from 'path';
import { IndexManifest } from '../manifest/index-manifest.model.js';
import { Manifest } from '../manifest/manifest.model.js';
import { ManifestService } from '../manifest/manifest.service.js';
import { FileBrowserService } from '../scanner/file-browser.service.js';
import { FileReaderService } from '../scanner/file-reader.service.js';
import { BackupOnClientService } from './backup-client.service.js';

describe('BackupOnClientService', () => {
  let service: BackupOnClientService;
  let manifestService: ManifestService;
  let fileReader: FileReaderService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [CoreModule],
      providers: [ManifestService, ProtobufService, FileReaderService, FileBrowserService, BackupOnClientService],
    }).compile();

    manifestService = module.get<ManifestService>(ManifestService);
    fileReader = module.get<FileReaderService>(FileReaderService);
    service = module.get<BackupOnClientService>(BackupOnClientService);
  });

  describe('#RefreshCache', () => {
    test('#RefreshCache should create one manifest from scratch ', async () => {
      let cnt = 0;
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries) => {
        cnt = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test2'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test3'), stats: { size: Long.fromNumber(300) } } };
        yield { fileManifest: { path: Buffer.from('/test4'), stats: { size: Long.fromNumber(400) } } };
        yield { fileManifest: { path: Buffer.from('/test5'), stats: { size: Long.fromNumber(500) } } };
        yield { fileManifest: { path: Buffer.from('/test6'), stats: { size: Long.fromNumber(600) } } };
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(1);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toBe(6);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });

    test('#RefreshCache should create two manifest from scratch ', async () => {
      const cnt: Record<string, number> = {};
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries, manifest: Manifest) => {
        cnt[manifest.journalPath] = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test2'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test3'), stats: { size: Long.fromNumber(300) } } };
        yield { fileManifest: { path: Buffer.from('/test4'), stats: { size: Long.fromNumber(400) } } };
        yield { fileManifest: { path: Buffer.from('/test5'), stats: { size: Long.fromNumber(500) } } };
        yield { fileManifest: { path: Buffer.from('/test6'), stats: { size: Long.fromNumber(600) } } };
        yield { header: { sharePath: Buffer.from('/etc') } };
        yield { fileManifest: { path: Buffer.from('/test32'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test7'), stats: { size: Long.fromNumber(100) } } };
        yield { fileManifest: { path: Buffer.from('/test8'), stats: { size: Long.fromNumber(200) } } };
        yield { fileManifest: { path: Buffer.from('/test9'), stats: { size: Long.fromNumber(300) } } };
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(2);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toMatchInlineSnapshot(`
        {
          "/tmp/backups.%2Fetc.journal": 4,
          "/tmp/backups.%2Fhome.journal": 6,
        }
      `);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });

    test('#RefreshCache should create error', async () => {
      const cnt: Record<string, number> = {};
      jest.spyOn(manifestService, 'deleteManifest').mockResolvedValue(undefined);
      jest.spyOn(manifestService, 'writeJournalEntry').mockImplementation(async (entries, manifest: Manifest) => {
        cnt[manifest.journalPath] = await count(entries);
      });

      function* myIterator() {
        yield { header: { sharePath: Buffer.from('/home') } };
        yield { fileManifest: { path: Buffer.from('/test1'), stats: { size: Long.fromNumber(100) } } };
        throw new NotFoundException('Should be return error');
      }

      const reply = await service.refreshCache(from(myIterator()));

      expect(manifestService.deleteManifest).toHaveBeenCalledTimes(1);
      expect(manifestService.deleteManifest).toHaveBeenCalledWith(new Manifest('backups.%2Fhome', '/tmp'));
      expect(cnt).toMatchInlineSnapshot(`{}`);
      expect(reply).toMatchSnapshot('RefreshCacheReply');
    });
  });

  describe('LaunchBackup', () => {
    let mockIndex: IndexManifest;

    beforeEach(() => {
      mockIndex = new IndexManifest();
      jest.spyOn(manifestService, 'loadIndex').mockImplementation(async (manifest: Manifest) => {
        expect(manifest.manifestPath).toBe('/tmp/backups.%2Fhome%2Fphoenix%2FDownloads%2Ftest.manifest');
        return mockIndex;
      });
      jest.spyOn(fileReader, 'getFiles').mockImplementation((index: IndexManifest) =>
        from(
          (function* fileReaderGetFiles(index: IndexManifest) {
            expect(index).toBe(mockIndex);
            const entry = index.getEntry(Buffer.from('/42093_X_Hot Rod.pdf'));
            if (entry) {
              entry.markViewed = true;
            }
            yield {
              chunks: [
                Buffer.from('782199a8de21390a804df2db8cc308c4d8966c10e181a252f96af35c51e9682f', 'hex'),
                Buffer.from('4a8602e2bc9f3eb9f744da31b0e1abf5474023e94a8bba97ecf544f578372fff', 'hex'),
              ],
              path: Buffer.from('/42093_X_Hot Rod.pdf'),
              sha256: Buffer.from('46b475c703c639f21c725daf1eed193ab1ea66a069028b512e0bca60533e037c', 'hex'),
              stats: {
                created: new Long(1271037455, 376),
                groupId: new Long(1000, 0),
                lastModified: new Long(1656647184, 371),
                lastRead: new Long(1063485888, 372),
                mode: new Long(33188, 0),
                ownerId: new Long(1000, 0),
                size: new Long(4503209, 0),
              },
            };
          })(index),
        ),
      );
    });

    test('should make backup', async () => {
      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [Buffer.from('*.pdf')],
        excludes: [Buffer.from('*.bak')],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });

    test('should make backup with index (remove elements)', async () => {
      mockIndex.add({
        path: Buffer.from('/42093_X_Hot Rod.pdf'),
        stats: {},
        acl: [],
        xattr: {},
        chunks: [],
      });
      mockIndex.add({
        path: Buffer.from('/fichier supprimÃ©.pdf'),
        stats: {},
        acl: [],
        xattr: {},
        chunks: [],
      });
      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [],
        excludes: [],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });

    test('but return an error', async () => {
      jest.spyOn(manifestService, 'loadIndex').mockRejectedValue(new Error("Can't read files"));

      const share = {
        sharePath: Buffer.from('/home/phoenix/Downloads/test'),
        includes: [],
        excludes: [],
      };
      const it = await toArray(service.launchBackup(share));
      expect(it).toMatchSnapshot('Launch backup (normal)');
    });
  });

  describe('#getChunk', () => {
    test('download a file', async () => {
      const testFile = join(__dirname, 'backup-client.service.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(0),
        size: Long.fromNumber(100000),
        sha256: Buffer.from('e4133d7518b34ba4ed12714c8cd2ace672696cfb5bef1fb3a2ad28f4492f4806', 'hex'),
        failIfWrongHash: true,
      });

      const allChunks = await toArray(chunks);

      expect(allChunks.map((c) => c.data.toString())).toMatchSnapshot('getChunk');
    });

    test('download a file, wrong hash', async () => {
      const testFile = join(__dirname, 'backup-client.service.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(0),
        size: Long.fromNumber(100000),
        sha256: Buffer.from('13', 'hex'),
        failIfWrongHash: true,
      });

      await expect(toArray(chunks)).rejects.toThrow(/The chunk .* should have a sha of 13, but is .*/);
    });

    test('download a file, piece of file', async () => {
      const testFile = join(__dirname, 'backup-client.service.ts');
      const chunks = service.getChunk({
        filename: Buffer.from(testFile),
        position: Long.fromNumber(10),
        size: Long.fromNumber(30),
        sha256: Buffer.from('fd4a6d2995cd3f227c79f77b7f51f5d59e032bac75a6d8e4905589925346633a', 'hex'),
        failIfWrongHash: true,
      });

      const allChunks = await toArray(chunks);

      expect(allChunks.map((c) => c.data.toString())).toMatchSnapshot('getChunk');
    });
  });
});
