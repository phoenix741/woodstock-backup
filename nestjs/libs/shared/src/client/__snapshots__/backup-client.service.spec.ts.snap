// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupOnClientService #RefreshCache #RefreshCache should create error: RefreshCacheReply 1`] = `
{
  "code": 1,
  "message": "Should be return error",
}
`;

exports[`BackupOnClientService #RefreshCache #RefreshCache should create one manifest from scratch : RefreshCacheReply 1`] = `
{
  "code": 0,
}
`;

exports[`BackupOnClientService #RefreshCache #RefreshCache should create two manifest from scratch : RefreshCacheReply 1`] = `
{
  "code": 0,
}
`;

exports[`BackupOnClientService #getChunk download a file, piece of file: getChunk 1`] = `
[
  "adRequestException, Injectable",
]
`;

exports[`BackupOnClientService #getChunk download a file: getChunk 1`] = `
[
  "import { BadRequestException, Injectable, Logger } from '@nestjs/common';
import { ExecuteCommandService, globStringToRegex, mangle, notUndefined, split } from '@woodstock/core';
import { createReadStream, ReadStream } from 'fs';
import { AsyncIterableX, concat, from, of } from 'ix/asynciterable';
import { catchError, concatAll, filter, finalize, map } from 'ix/asynciterable/operators';
import { Manifest, ManifestService } from '../manifest';
import {
  ChunkInformation,
  ExecuteCommandReply,
  FileChunk,
  FileManifest,
  LaunchBackupReply,
  RefreshCacheReply,
  RefreshCacheRequest,
  Share,
  StatusCode,
} from '../protobuf/woodstock.interface';
import { FileHashReader, FileReaderService } from '../scanner';

@Injectable()
export class BackupOnClientService {
  private logger = new Logger(BackupOnClientService.name);

  constructor(
    private commandService: ExecuteCommandService,
    private fileReader: FileReaderService,
    private manifestService: ManifestService,
  ) {}

  async executeCommand(command: string): Promise<ExecuteCommandReply> {
    return this.commandService.executeCommand(command);
  }

  private async createManifestForSharePath(sharePath: Buffer, source: AsyncIterable<FileManifest | undefined>) {
    const manifest = new Manifest(\`backups.\${mangle(sharePath)}\`, '/tmp/');
    await this.manifestService.deleteManifest(manifest);

    const entries$ = from(source).pipe(
      notUndefined<FileManifest>(),
      map((fileManifest) => ManifestService.toAddJournalEntry(fileManifest, true)),
    );
    await this.manifestService.writeJournalEntry(entries$, manifest);

    await this.manifestService.compact(manifest);
  }

  private async createManifestsFromSource(request: AsyncIterableX<RefreshCacheRequest>) {
    this.logger.log(\`Start creating manifests from source\`);

    const isHeader = (v: RefreshCacheRequest) => v.header?.sharePath;
    const groupOfManifest = request.pipe(split(isHeader));

    for await (const group of groupOfManifest) {
      const sharePath = group.key;
      const manifests = group.iterable.pipe(
        map((v) => v.fileManifest),
        notUndefined(),
      );

      await this.createManifestForSharePath(sharePath, manifests);
    }
  }

  async refreshCache(request: AsyncIterableX<RefreshCacheRequest>): Promise<RefreshCacheReply> {
    try {
      await this.createManifestsFromSource(request);

      return { code: StatusCode.Ok };
    } catch (err) {
      return { code: StatusCode.Failed, message: err.message };
    }
  }

  launchBackup(share: Share): AsyncIterableX<LaunchBackupReply> {
    const manifest = new Manifest(\`backups.\${mangle(share.sharePath)}\`, '/tmp/');

    const createIndex$ = from(this.manifestService.loadIndex(manifest));

    const addEntryToIndex$ = createIndex$.pipe(
      map((index) =>
        this.fileReader.getFiles(
          index,
          share.sharePath,
          share.includes?.map((s) => globStringToRegex(s.toString('latin1'))),
          share.excludes?.map((s) => globStringToRegex(s.toString('latin1'))),
        ),
      ),
      concatAll(),
      map((entry) => ({ entry, response: undefined })),
    );

    const addRemoveToIndex$ = createIndex$.pipe(
      map((index) => from(index.walk())),
      concatAll(),
      filter((entry) => !entry.markViewed),
      map((file) => ManifestService.toRemoveJournalEntry(file.path)),
      map((entry) => ({ entry, response: undefined })),
    );

    const entries$ = concat(
      addEntryToIndex$,
      addRemoveToIndex$,
      of({ entry: undefined, response: { code: StatusCode.Ok } }),
    );

    return entries$.pipe(
      catchError((err) => {
        return of({
          entry: undefined,
          response: { code: StatusCode.Failed, message: err.message },
        });
      }),
    );
  }

  getChunk(request: ChunkInformation & { failIfWrongHash?: boolean }): AsyncIterableX<FileChunk> {
    const { filename, position, size, sha256 } = request;
    this.logger.debug(\`Get the chunk \${filename} at position \${position} (size: \${size})\`);

    const stream: ReadStream = createReadStream(filename, {
      start: position.toNumber(),
      end: position.add(size).sub(1).toNumber(),
    });
    const hashReader = new FileHashReader();

    stream.pipe(hashReader);
    stream.on('error', (err) => {
      hashReader.emit('error', err);
      hashReader.end();
      stream.close();
    });
    hashReader.on('error', () => {
      hashReader.end();
      stream.close();
    });

    return from(hashReader).pipe(
      map((data) => ({ data })),
      finalize(() => {
        if (request.failIfWrongHash && sha256 && hashReader.hash && !hashReader.hash.equals(sha256)) {
          const message = \`The chunk \${filename.toString()}:\${position}:\${size} should have a sha of \${sha256.toString(
            'hex',
          )}, but is \${hashReader.hash.toString('hex')}\`;
          this.logger.warn(message);

          throw new BadRequestException(message);
        }
      }),
    );
  }
}
",
]
`;

exports[`BackupOnClientService LaunchBackup but return an error: Launch backup (normal) 1`] = `
[
  {
    "entry": undefined,
    "response": {
      "code": 1,
      "message": "Can't read files",
    },
  },
]
`;

exports[`BackupOnClientService LaunchBackup should make backup with index (remove elements): Launch backup (normal) 1`] = `
[
  {
    "entry": {
      "chunks": [
        {
          "data": [
            120,
            33,
            153,
            168,
            222,
            33,
            57,
            10,
            128,
            77,
            242,
            219,
            140,
            195,
            8,
            196,
            216,
            150,
            108,
            16,
            225,
            129,
            162,
            82,
            249,
            106,
            243,
            92,
            81,
            233,
            104,
            47,
          ],
          "type": "Buffer",
        },
        {
          "data": [
            74,
            134,
            2,
            226,
            188,
            159,
            62,
            185,
            247,
            68,
            218,
            49,
            176,
            225,
            171,
            245,
            71,
            64,
            35,
            233,
            74,
            139,
            186,
            151,
            236,
            245,
            68,
            245,
            120,
            55,
            47,
            255,
          ],
          "type": "Buffer",
        },
      ],
      "path": {
        "data": [
          47,
          52,
          50,
          48,
          57,
          51,
          95,
          88,
          95,
          72,
          111,
          116,
          32,
          82,
          111,
          100,
          46,
          112,
          100,
          102,
        ],
        "type": "Buffer",
      },
      "sha256": {
        "data": [
          70,
          180,
          117,
          199,
          3,
          198,
          57,
          242,
          28,
          114,
          93,
          175,
          30,
          237,
          25,
          58,
          177,
          234,
          102,
          160,
          105,
          2,
          139,
          81,
          46,
          11,
          202,
          96,
          83,
          62,
          3,
          124,
        ],
        "type": "Buffer",
      },
      "stats": {
        "created": Long {
          "high": 376,
          "low": 1271037455,
          "unsigned": false,
        },
        "groupId": Long {
          "high": 0,
          "low": 1000,
          "unsigned": false,
        },
        "lastModified": Long {
          "high": 371,
          "low": 1656647184,
          "unsigned": false,
        },
        "lastRead": Long {
          "high": 372,
          "low": 1063485888,
          "unsigned": false,
        },
        "mode": Long {
          "high": 0,
          "low": 33188,
          "unsigned": false,
        },
        "ownerId": Long {
          "high": 0,
          "low": 1000,
          "unsigned": false,
        },
        "size": Long {
          "high": 0,
          "low": 4503209,
          "unsigned": false,
        },
      },
    },
    "response": undefined,
  },
  {
    "entry": {
      "manifest": {
        "acl": [],
        "chunks": [],
        "path": {
          "data": [
            47,
            102,
            105,
            99,
            104,
            105,
            101,
            114,
            32,
            115,
            117,
            112,
            112,
            114,
            105,
            109,
            195,
            169,
            46,
            112,
            100,
            102,
          ],
          "type": "Buffer",
        },
        "stats": undefined,
        "xattr": {},
      },
      "type": 2,
    },
    "response": undefined,
  },
  {
    "entry": undefined,
    "response": {
      "code": 0,
    },
  },
]
`;

exports[`BackupOnClientService LaunchBackup should make backup: Launch backup (normal) 1`] = `
[
  {
    "entry": {
      "chunks": [
        {
          "data": [
            120,
            33,
            153,
            168,
            222,
            33,
            57,
            10,
            128,
            77,
            242,
            219,
            140,
            195,
            8,
            196,
            216,
            150,
            108,
            16,
            225,
            129,
            162,
            82,
            249,
            106,
            243,
            92,
            81,
            233,
            104,
            47,
          ],
          "type": "Buffer",
        },
        {
          "data": [
            74,
            134,
            2,
            226,
            188,
            159,
            62,
            185,
            247,
            68,
            218,
            49,
            176,
            225,
            171,
            245,
            71,
            64,
            35,
            233,
            74,
            139,
            186,
            151,
            236,
            245,
            68,
            245,
            120,
            55,
            47,
            255,
          ],
          "type": "Buffer",
        },
      ],
      "path": {
        "data": [
          47,
          52,
          50,
          48,
          57,
          51,
          95,
          88,
          95,
          72,
          111,
          116,
          32,
          82,
          111,
          100,
          46,
          112,
          100,
          102,
        ],
        "type": "Buffer",
      },
      "sha256": {
        "data": [
          70,
          180,
          117,
          199,
          3,
          198,
          57,
          242,
          28,
          114,
          93,
          175,
          30,
          237,
          25,
          58,
          177,
          234,
          102,
          160,
          105,
          2,
          139,
          81,
          46,
          11,
          202,
          96,
          83,
          62,
          3,
          124,
        ],
        "type": "Buffer",
      },
      "stats": {
        "created": Long {
          "high": 376,
          "low": 1271037455,
          "unsigned": false,
        },
        "groupId": Long {
          "high": 0,
          "low": 1000,
          "unsigned": false,
        },
        "lastModified": Long {
          "high": 371,
          "low": 1656647184,
          "unsigned": false,
        },
        "lastRead": Long {
          "high": 372,
          "low": 1063485888,
          "unsigned": false,
        },
        "mode": Long {
          "high": 0,
          "low": 33188,
          "unsigned": false,
        },
        "ownerId": Long {
          "high": 0,
          "low": 1000,
          "unsigned": false,
        },
        "size": Long {
          "high": 0,
          "low": 4503209,
          "unsigned": false,
        },
      },
    },
    "response": undefined,
  },
  {
    "entry": undefined,
    "response": {
      "code": 0,
    },
  },
]
`;
